### PR TITLE
feat: implement Flight and gRPC services for RegionServer

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement a new flight get that only accepts the logical plan. And the `RegionServer` gRPC services for region server struct. Add a new temporary ENV `ENABLE_REGION_SERVER` to control whether to start the region server for datanode.

This patch also re-organizes the traits around `Server` and `GrpcServer`:
- change all the service handlers to optional
- Extract a `FlightCraft` as the handler trait for flight services, to reduce the redundant "unimplemented".
- Define `RegionServerHandler` for region server requests

For the sake of clarity, refactor to `GreptimeDatabase` and `GreptimeRequestHandler` are not included in this PR.

After this refactoring, their relationship will become something like the following. Firstly, all related trait are defined in `servers`:
```mermaid
erDiagram
    Server {
        HttpServer HTTP
        GrpcServer gRPC
        others others
    }
    Server ||--|{ GrpcServer : contains
    GrpcServer {
        optional Flight
        optional RegionServer
        optional Database
        option HealthCheck
        optional PromQuery
    }
```
And each service (`Database` or `Flight`) has one trait to define the required methods. E.g., `FlightCraft` for Flight services
```rust
pub trait FlightCraft: Send + Sync + 'static {
    async fn do_get(
        &self,
        request: Request<Ticket>,
    ) -> TonicResult<Response<TonicStream<FlightData>>>;
}
```

`GrpcServer` only requires those trait objects, and will build corresponding services by them. 

These are all in the `servers` sub-crate. And other implementations (`datanode` and `frontend`) need to 
- implement the required trait for handlers. Like `impl RegionServerHandler for RegionServer`
- construct the server struct corresponding to the protocol, like `GrpcServer` for `gRPC`
- start the server

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Review by MichaelScofield — APPROVED** [review]

File: src/servers/src/grpc/region_server.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+#[async_trait]
+pub trait RegionServerHandler: Send + Sync {
+    async fn handle(&self, request: RequestBody) -> Result<RegionResponse>;
+}
+
+pub type RegionServerHandlerRef = Arc<dyn RegionServerHandler>;
+
+#[derive(Clone)]
+pub struct RegionServerRequestHandler {
```
Comment (evenyag) [inline src/servers/src/grpc/region_server.rs:51]: question(non-blocking): We already define `RegionServerHandler`. Could we rename RegionServerRequestHandler to `RegionRequestHandler`? It handles RegionRequest.

Or rename `RegionServerHandler` to `RequestBodyHandler`...

We can also rename it to `RegionServerServiceImpl`...

**Review by evenyag — APPROVED** [review]
Since all comments are non-blocking, I'd like to approve this.

File: src/servers/src/grpc/greptime_handler.rs [hunk 2]
```diff
@@ -174,7 +177,7 @@ pub(crate) fn create_query_context(header: Option<&RequestHeader>) -> QueryConte
 /// Histogram timer for handling gRPC request.
 ///
 /// The timer records the elapsed time with [StatusCode::Success] on drop.
-struct RequestTimer {
+pub(crate) struct RequestTimer {
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/servers/src/grpc/greptime_handler.rs:180]: they are unrelated scopes. Method's is narrower than struct's

File: src/servers/src/grpc/region_server.rs (hunk 1 above)
Comment (waynexia) [reply src/servers/src/grpc/region_server.rs:51]: Going to rename in the follow-up pr

**Pull request merged**